File: combination-sum-iv/combination-sum-iv.cpp
class Solution {
public:
     int combinationSum4(vector<int>& nums, int target) {
        unsigned int t[target+1];
        memset(t,0,sizeof(t));
        t[0] = 1;
        for (int i = 1; i <= target; ++i) {
            for (int j=1; j<=nums.size();j++) {
                if (i >=nums[j-1]) 
                    t[i] += t[i - nums[j-1]];
            }
        }
        
        return t[target];
    }
    
};

// class Solution {
// public:
//      int combinationSum4(vector<int>& nums, int target) {
//         vector<unsigned int> result(target + 1);
//         result[0] = 1;
//         for (int i = 1; i <= target; ++i) {
//             for (int x : nums) {
//                 if (i >= x) result[i] += result[i - x];
//             }
//         }
        
//         return result[target];
//     }
    
// };

File: 377-combination-sum-iv/combination-sum-iv.py
from typing import List

class Solution:
    def combinationSum4(self, nums: List[int], target: int) -> int:
        dp = [None] * (target + 1)
        
        def backtrack(current_sum: int) -> int:
            if current_sum == target:
                return 1
            if current_sum > target:
                return 0
            if dp[current_sum] is not None:
                return dp[current_sum]
            
            total_combinations = 0
            for num in nums:
                total_combinations += backtrack(current_sum + num)
            
            dp[current_sum] = total_combinations
            return dp[current_sum]
        
        return backtrack(0)
